### ISSUE TITLE
iR+ Process for gene naming and analysis

### ISSUE THREAD

**Issue by bcorrie** [ireceptor-plus/issues#58]
In the recent SciTech meeting the topic around the family/subgroup -> gene -> allele mapping, relevant to Stats and other analysis was brought up. It would be beneficial for the iR+ groups to share their experiences as to how this is done...

**Comment by bcorrie** [comment]
Start off with iReceptor - current status/approach:

- Assume IGMT nomenclature
- Parse allele name with regex and derive gene and family
- Store in the database with an index at data load time, so can do very fast, exact match searches across billions of records
- Repository uses internal representation to count and store stats
- Stats API returns internally cached stats

**Comment by bcorrie** [comment]
> * Parse allele name with regex and derive gene and family

We know we are not doing this correctly at the moment - too many edge cases - regex doesn't really work... 8-)
This breaks down for other species - See #8

**Comment by bussec** [comment]
@bcorrie continuing part of our [dataloader #44](https://github.com/sfu-ireceptor/dataloading-mongo/issues/44) discussion here:

In sciReptor we basically gave up on trying to parse gene symbols, we are using the strings as identifiers to look up that information in pre-compiled table. But this is not ideal either, as it still sticks uses something that is a label as an identifier.

I think to handle the complex genetics of the involved loci, we need to replace gene symbols with PIDs for most of the annotation, storage and retrieval part. The current symbols (labels) for an PID then need to be resolved during the user query. The database used for this lookup needs to be shared among repositories, similar to the ontologies we are using for other fields. Any chance this could work in real life?

**Comment by bcorrie** [comment]
Yes, I think so... We need a PID system though and a database to look those PIDs up in - and my understanding is hopefully that is OGRDB 8-) Until then we are stuck with what we have... 

If you think of our current v_call, I could see this being a PID or Ontology field. `v_call.id = ALLELE:42`, `v_call.label = IGHV1-69*01`

Our ontology/CURIE mechanism can be used to map this to a precise definition in a specific repository.

ALLELE IRI = http://ogrdb.org/allele/
ALLELE IRI = http://imgt.org/genedb/

So I can look up `ALLELE:42` with http://ogrdb.org/allele/42 or http://imgt.org/genedb/42 (or something along those lines). The important thing is the `ALLELE:42` resolves to the same object in the two databases, and IDEALLY the `label` that is returned from both is the canonical `IGHV1-69*01`

Presumably you would be able to look up `ALLELE:42` and get its locus, family, gene, and allele info as well as the `label`. This gives you precise, comparable, and computable data. Going from PID to data is easy...

The really big question for me, is how does the user go the other way. My understanding is that people think in the terms of the labels... So how does a user state that the want all of the `IGHV1-69*01` v_call's? They aren't going to know to ask for `ALLELE:42` 8-) The user conceptually uses the labels, so we need labels to work and be consistent. That is why around our OGRDB discussions I am a bit pedantic about "canonical name" or "canonical label". "IGHV1-69*01" needs to be like "Homo sapiens" - there needs to be an official label - you can have synonyms (e.g. V1-69-01, human) but the PID/Ontology defines the "canonical label"

It gets worse I think... How are you going to ask for all of the IGHV1-69 genes? So you probably need PIDs for each of the levels: locus, family, gene, and allele? Or, you need an actual Ontology, which encapsulates this structure.

This structure/hierarchy is what we calculate now, based on the only thing we have, the gene call. In the Turnkey, we force it such that when data is loaded, it is interoperable and reusable. Which means you have to use the right nomenclature OR we write a specific data loader for a tool that produces a messed up nomenclature and convert it. We enforce this, even though this isn't part of the AIRR Standard.

This is what we did for Adaptive, we have a custom data loader that loads Adaptive gene calls and converts them to IMGT nomenclature, then runs that through our gene hierarchy building (locus, family, gene, allele) 8-)

**Comment by schristley** [comment]
> Yes, I think so... We need a PID system though and a database to look those PIDs up in - and my understanding is hopefully that is OGRDB 8-) Until then we are stuck with what we have...
> 
> If you think of our current v_call, I could see this being a PID or Ontology field. `v_call.id = ALLELE:42`, `v_call.label = IGHV1-69*01`
> 
> Our ontology/CURIE mechanism can be used to map this to a precise definition in a specific repository.

I agree we need a PID system too, and the CURIE implementation is nice as it also can be "lifted" into a decentralized identifier, which is just what we want for OGRDB. However, don't limit ourselves to ontologies only, and the label is optional. With a decentralized identifier, think of ALLELE as a CONTROLLER that knows how to interpret the value of 42, but it still operates the same way in generating an IRI.

> The really big question for me, is how does the user go the other way. My understanding is that people think in the terms of the labels... So how does a user state that the want all of the `IGHV1-69*01` v_call's?

In the standard way I think, query on a GeneDescription API, searching the label field for `IGHV1-69*01` to get the current identifier, ALLELE:42, then use that identifier to search. Presumably, at any given time, a gene label will be associated with a single identifier.

> It gets worse I think... How are you going to ask for all of the IGHV1-69 genes? So you probably need PIDs for each of the levels: locus, family, gene, and allele? Or, you need an actual Ontology, which encapsulates this structure.

There's only one sequence, i.e. the germline sequence used for the annotation, and that has an identifier. Attached to that identifier are all of those metadata fields you mention: locus, family, gene, and allele. That sequence typically equates to the allele level, though. Yes, there is biological knowledge encoded there. I don't see that being problematic as this knowledge is (mostly) well-known.

**Comment by schristley** [comment]
> I think to handle the complex genetics of the involved loci, we need to replace gene symbols with PIDs for most of the annotation, storage and retrieval part. The current symbols (labels) for an PID then need to be resolved during the user query. The database used for this lookup needs to be shared among repositories, similar to the ontologies we are using for other fields. Any chance this could work in real life?

Yep it will work, though the database does not need to be shared, as it is a decentralized database. However, yes, a significant amount of data is shared among all those decentralized databases. This database is small, having your own local copy isn't a big deal, as you will have it anyways (at least in file form) if you are doing any sequence annotation. Here's another way to think about it, the database is so small that it naturally becomes decentralized, everybody has a copy. Furthermore, it naturally follows the data it is attached to, when I send you an AIRR data set, this data might come with the repertoire metadata for local search...

Also, we need to think about the progression of the current software and data infrastructure to the new, which leaves as much flexibility up to the repository that we can give. The decentralized design gives a lot of flexibility to individual repositories for how they want to handle storage and retrieval. We also don't want to force analysis tools to perform HTTP lookups, so a local database with local search is a MUST provide. Defining the file format will cover the local database and tools can perform their own search over this data.

@bussec I think the biggest decision for you is if you want to manage your own germline database and (importantly) generate your own IDs. Here there are two IDs, generating new germline set identifiers and generating new allele sequence identifiers. If you do, then you need to generate them as a decentralized identifier. I'm assuming that VDJServer will eventually do both.

The next decision is if you want to run a query service for your germline database, i.e. respond to HTTP requests for data with your generated IDs. I don't think that this should be mandatory. If you are running a repository, then you might because you have the infrastructure. If you are doing a local analysis, then you have a local database in a file. The database is simple enough that clients can perform their own searching, so the only thing that is mandatory is the ability to acquire the database. If a repository does not want to provide a query service, there just needs to be a simple mechanism to download the database so local search can be performed.

**Comment by bcorrie** [comment]
> > The really big question for me, is how does the user go the other way. My understanding is that people think in the terms of the labels... So how does a user state that the want all of the `IGHV1-69*01` v_call's?
> 
> In the standard way I think, query on a GeneDescription API, searching the label field for `IGHV1-69*01` to get the current identifier, ALLELE:42, then use that identifier to search. 

That is my main point. This only works if there is a "canonical" value for the label field that is searchable. By canonical, I mean "conforming to a general rule or acceptable procedure" - that is the label is generally accepted by the community and therefore when the user thinks they are looking for something there is a community accepted (canonical) way to ask for it (e.g. `IGHV1-69*01`). 

This of course requires v_call to NOT be a string (e.g. `IGHV1-69*01`) but instead be a PID (e.g. ALLELE:42) (possibly with a label).  Or more likely to maintain backwards compatibility, we might have a new field called v_call_id.

> Presumably, at any given time, a gene label will be associated with a single identifier.

This is the part that worries me... This "should" be true within a specific germline database, but you can't expect that in general (see below)... And if you can't expect it in general how can you compare two data sets. For example:

1. - @bussec is maintaining a germline database, has a Gene ID (ID1) for Gene G1, and it has Label L. @schristley is maintaining a germline database, has a Gene ID (ID1) for Gene G1, and it has Label L. Genes are the same, Gene IDs are the same Label is the same. All is good, this is the normal, canonical world 8-)
2. - @bussec is maintaining a germline database, finds a new Gene (G10) and creates a new ID (ID10) for that Gene, and gives it Label L10. @schristley is maintaining a different germline database, finds a new different Gene (G11) creates a different ID (ID11) for a for the Gene but gives it Label L10. Genes are different, Gene IDs are different, but the labels are the same. 
3. - @bussec is maintaining a germline database, finds a new Gene (G20) creates a new ID (ID20) for that Gene, and labels it L20. @schristley doesn't know about @bussec, finds the same new Gene (G20) creates a new ID (ID21) for Gene G20, and labels it L21. Genes are the same, Gene IDs are different, and the Labels are different.

How do I compare data sets generated by @bussec and @schristley? I download their data sets, I download their germline definitions, and then I scratch my head... They describe the individual data sets just fine, but I am stuck on comparing the datasets unless I essentially do a full cross comparison on the two germline sets to figure out which IDs maps to which Genes. Which kind of makes a germline database not very useful for comparing data sets 8-)

Now maybe that is what we need to handle the current state of affairs, but that isn't very pretty!

**Comment by bussec** [comment]
@bcorrie My two cents on those scenarios:
* Scenario 1 is basically what we have in the ontology world. An ontology is curated by an authority who shares updated versions via its users. Users will have local copies, but only use them in a read-only fashion.
* If `label` and `id` are assigned at the same time, then the only thing we might solve, is the problem of parsing (via an `id` indexed look-up table). But this is not what I am thinking about: I would like to use `id` as a quick and decentralized way to make gene records usable for annotation without longer delays. The labels will take longer to be assigned and this is not a problem if our tools can work also with only-`id` records.
* I do **not** expect that `labels` will be assigned by the germline database but rather by a committee similar to IARC. Otherwise it will be difficult to avoid collisions as you described them in your scenarios 2+3.

**Comment by schristley** [comment]
> How do I compare data sets generated by @bussec and @schristley?

I don't think the reality will be overly extreme. If we are talking human for example, most of the genes will be identical between the two germlines with the same ids, label, annotation, etc., and thus are directly comparable. For a new gene, there are processes in place to get that new gene recognized, give it a real label, etc. The researcher can assign it a local id and label, but they can also ask OGRDB for an AIRR id and label.

When and if that new gene is approved, given an "official" id and label, and added to the set of curated genes in the AIRR germline sets, then we need to resolve the duplicative IDs, if the researcher assigned a local ID. If they used the AIRR id and label then they are set. The plan to handle that is the same as plan for handling IMGT's name changing, in particular there will be an alias field(s) for each gene that contains prior IDs and labels used for the gene. Handling aliases is a pain in the ass, but unfortunately, gene aliases have become ubiquitous throughout bioinformatics.

**Comment by bcorrie** [comment]
> I don't think the reality will be overly extreme. If we are talking human for example, most of the genes will be identical between the two germlines with the same ids, label, annotation, etc., and thus are directly comparable. For a new gene, there are processes in place to get that new gene recognized, give it a real label, etc. The researcher can assign it a local id and label, but they can also ask OGRDB for an AIRR id and label.
> 
> When and if that new gene is approved, given an "official" id and label, and added to the set of curated genes in the AIRR germline sets, then we need to resolve the duplicative IDs, if the researcher assigned a local ID. If they used the AIRR id and label then they are set. The plan to handle that is the same as plan for handling IMGT's name changing, in particular there will be an alias field(s) for each gene that contains prior IDs and labels used for the gene. Handling aliases is a pain in the ass, but unfortunately, gene aliases have become ubiquitous throughout bioinformatics.

I think if this is how things roll out, then I agree, we are fine.  That is about as good as you can expect. The standard should have "strong" wording that this is what should happen.

This is saying what I was saying above, that there is general a "canonical" label for each gene (in general controlled by a process and accepted by the community). And that when you search for a ID for a gene in a germline database (any germline database) you get essentially the same information back (including the same "canonical" label). And vice versa, when you search for a "canonical" label, you get the same Gene ID back. You can search for synonyms (aliases), sure, but if you use the canonical label, then you get an expected and deterministic results. Searching for aliases, your results may vary, depending on the germline database - I expect that not all germline databases would have the same synonyms for the same Gene ID.

My concern is the above won't happen - primarily based on our past discussion in the AIRR Community around unwillingness to be specific and precise around describing what should go, as a text string, into v_call 8-) To me this is the canonical name for a Gene, and our description of this field in the AIRR spec is very imprecise. In many of our discussions around germline and v/d/j/c_call, there has been a fair bit of pushback (at least in my interpretation) around requiring a "canonical" gene naming. You can see this in the way we describe the v_call (and d/j/c_call) fields in the spec, pretty well anything goes (ok, I am being provocative a bit 8-)

The reason we "force" a gene naming nomenclature in the Turnkey data loader that is more "FAIR" (or at least "IR") that the AIRR spec requires is because if we used the AIRR spec and loaded valid AIRR TSV files as they are, the data in two Turnkey repositories (and in general data in the AIRR Data Commons) would not be interoperable and reusable - which is what started this conversation in the first place 8-)

**Comment by bussec** [comment]
@bcorrie When you are using `ALLELE:42` as an ID example, would you expect that all database use `ALLELE` as a suffix (potientially resolving it to different IRIs internally) or could there also be an `ADCGERMLINE:42`, `VDJBASE:314159` and `TUMBLEWEED:23`?

**Comment by bcorrie** [comment]
> @bcorrie When you are using ALLELE:42 as an ID example, would you expect that all database use ALLELE as a suffix (potientially resolving it to different IRIs internally) or could there also be an ADCGERMLINE:42, VDJBASE:314159 and TUMBLEWEED:23?

Not sure. I think ideally `ALLELE:42` would be THE identifier that points to THAT germline sequence and its various gene metadata fields (label, locus, gene, family, allele...). As Scott points out, there is only one germline sequence that takes on a specific set of NTs, so it isn't hard to imagine that one could come up with an identifier for it. `ALLELE` could be replaced by a various IRI with associated providers, but just like 9606 identifies one thing and one thing only in the NCBI Taxonomy, 42 would identify one germline sequence, and one germline sequence only. I think there could be a number of IRI and providers for `ALLELE:42` and one could have IRI prefixes point to http://germline.adcgermling.org/allele/ and http://ogrdb.org/ogrdb/v2/allele?id= and http://gbase.tubleweed.com/life/universe/everything/

But we need 42 to identify the `thing` - in this case the specific sequence we are talking about and for the above to resolve to returning information about that thing...

Without this data in the ADC will not be interoperable and comparable - or at least not practically.

I think it would be a disaster if ADCGERMLINE:42, VDJBASE:314159 and TUMBLEWEED:23 all were IDs that represented the same germline sequence as `ALLELE:42`. Not sure if that is what you were asking. Imagine if they were used in different AIRR-seq repositories. How would I even determine that VDJBASE:314159 and TUMBLEWEED:23 were talking about the same thing, let alone figure out how to ask the repositories if they contained the equivalent of `ALLELE:42`

It would be less of a disaster if ADCGERMLINE:42 and ALLELE:42 represented the same thing, as the `42` is the ID - but I think this is redundant if you have IRI prefixes.

**Comment by bcorrie** [comment]
Bottom line, I believe it should be a PID at a minimum, if not an ontology. Recall our recommendation for PIDs in the AIRR OntoVoc is that the IRI prefix for a PID field is fixed by the spec.

**Comment by bussec** [comment]
> there is only one germline sequence that takes on a specific set of NTs, so it isn't hard to imagine that one could come up with an identifier for it

Not sure whether we are on the same page here. I agree that not having multiple records (i.e. with different IRIs) for the same allele is clearly the desired state. However, I am not sure that a decentralized system will be in this state all the time. One of the key problems is that:

* every allele has at any time point exactly one reference sequence associated with it, **but**
* multiple alleles may have identical reference sequences, typically due to a recent genomic duplication in the locus (e.g. mouse [_Ighv1-62-2_](https://www.ensembl.org/Mus_musculus/Transcript/Exons?t=ENSMUST00000194581) and [_Ighv1-71_](https://www.ensembl.org/Mus_musculus/Transcript/Exons?t=ENSMUST00000196690)).

Therefore you cannot derive *local ID* that would be consistent across nodes for the sequence alone, neither by hashing nor by a centralized "oracle". The biological information to resolve this -- the physical location on the chromosome -- will in most cases not be available.

This does not argue against requiring people to check the current database before minting a new ID, but it is a classical "write cache" issue, so duplicates may arise and need to be resolve at a later point.

**Comment by bussec** [comment]
But coming back to the various CURIEs: The main point was whether there can be more than one provider for non-overlapping alleles. So that e.g. `ADCGERMLINE` would point to the core database that only contains genomically mapped alleles, while `ALLELE` contains stuff approved by IARC and `TUMBLEWEED` automatically produced nightly builds for VDJbase.

**Comment by bussec** [comment]
Last point: While every allele has at any time point exactly one reference sequence associated with it, this does not mean that this sequence is always the same. Of course it MUST NOT change in any arbitrary fashion, but a reference sequence MAY be superseeded by a longer reference sequence of which it is a substring and which covers it completely. We can handle this e.g. by versioning the records, but it is important that an older gene call annotation still points to the then-recent version of the record.

**Comment by schristley** [comment]
> * multiple alleles may have identical reference sequences, typically due to a recent genomic duplication in the locus (e.g. mouse [_Ighv1-62-2_](https://www.ensembl.org/Mus_musculus/Transcript/Exons?t=ENSMUST00000194581) and [_Ighv1-71_](https://www.ensembl.org/Mus_musculus/Transcript/Exons?t=ENSMUST00000196690)).

That's a good point. We might be able to add some salt to the hash function so that different IDs will be generated.

**Comment by bcorrie** [comment]
> Not sure whether we are on the same page here. I agree that not having multiple records (i.e. with different IRIs) for the same allele is clearly the desired state. However, I am not sure that a decentralized system will be in this state all the time. One of the key problems is that:
> 
> * every allele has at any time point exactly one reference sequence associated with it, **but**
> * multiple alleles may have identical reference sequences, typically due to a recent genomic duplication in the locus (e.g. mouse [_Ighv1-62-2_](https://www.ensembl.org/Mus_musculus/Transcript/Exons?t=ENSMUST00000194581) and [_Ighv1-71_](https://www.ensembl.org/Mus_musculus/Transcript/Exons?t=ENSMUST00000196690)).
> 
> Therefore you cannot derive _local ID_ that would be consistent across nodes for the sequence alone, neither by hashing nor by a centralized "oracle". The biological information to resolve this -- the physical location on the chromosome -- will in most cases not be available.

OK, being a non-biologist I am trying to get down to what is the unique `thing` we are trying to identify. That `thing` should have a unique ID, no?

I thought that was the germline sequence, but you are saying that is not the case. So it sounds like it is the germline sequence combined with the physical location on the chromosome. Don't you want a unique ID for that `thing` that is the same in all databases and by all providers???

So `Ighv1-62-2` and `Ighv1-71` have the same sequence but are located at different locations on the chromosome. Biologically they are two different things, so they have two different IDs. But they seem quite concrete to me - one ID refers to sequence S at location N1 on chromosome C1 while the second ID refers to sequence S at location N2 on chromosome C2. I am sure I am missing something 8-)

**Comment by bcorrie** [comment]
> But coming back to the various CURIEs: The main point was whether there can be more than one provider for non-overlapping alleles. So that e.g. `ADCGERMLINE` would point to the core database that only contains genomically mapped alleles, while `ALLELE` contains stuff approved by IARC and `TUMBLEWEED` automatically produced nightly builds for VDJbase.

But you don't need CURIE prefixes for this do you??? You need different providers, but not CURIE prefixes. I am confused when you are using the ALL CAPS notation to refer to providers. The ALL CAPS format in the CURIE context is a CURIE prefix. Providers and CURIE prefixes are very different things. A single CURIE prefix (`ALLELE`) can have many providers. And those providers can behave as you describe above.

The "ADC Germline *Provider*" has IDs for only genomically mapped alleles, while the "OGRDB *Provider*" has the same IDs but in addition has alleles that are approved by IARC, and finally the "VDJBase *Provider*" might have a nightly build of everything in the "OGRDB *Provider*" plus a bunch of new stuff from VDJBase. Each *Provider* could have different sets of germline gene allele IDs.

But they all use the same CURIE format of `prefix:ID` with a given AIRR PID/Ontology field having a single CURIE `prefix` (e.g. `ALLELE`). When *ANY* one of those providers is talking about a specific `thing` (e.g. a specific sequence at a specific location on a specific chromosome) they *ALL* use the same ID.

If there are new `things`, then the new `things` need to have temporary IDs (and there are mechanisms to do things like that), but the agreed on `things` should have consistent IDs... And maybe we need mechanisms to have different classes of IDs (accepted, under consideration, and experimental) to provide some ID flexibility, but that shouldn't have a negative impact on the processing of things that are common across *Providers*.
